homeassistant/components/mqtt/update.py

```diff
@@ -0,0 +1,199 @@
+
+from collections.abc import Callable
+import functools
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components import update
+from homeassistant.components.update import (
+    DEVICE_CLASSES_SCHEMA,
+    UpdateEntity,
+    UpdateEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_DEVICE_CLASS, CONF_NAME, CONF_VALUE_TEMPLATE
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.typing import ConfigType
+
+from . import subscription
+from .config import DEFAULT_RETAIN, MQTT_RO_SCHEMA
+from .const import (
+    CONF_COMMAND_TOPIC,
+    CONF_ENCODING,
+    CONF_QOS,
+    CONF_RETAIN,
+    CONF_STATE_TOPIC,
+)
+from .debug_info import log_messages
+from .mixins import MQTT_ENTITY_COMMON_SCHEMA, MqttEntity, async_setup_entry_helper
+from .models import MqttValueTemplate, ReceiveMessage
+from .util import get_mqtt_data, valid_publish_topic, valid_subscribe_topic
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "MQTT Update"
+
+CONF_LATEST_VERSION_TEMPLATE = "latest_version_template"
+CONF_LATEST_VERSION_TOPIC = "latest_version_topic"
+CONF_PAYLOAD_INSTALL = "payload_install"
+
+
+PLATFORM_SCHEMA_MODERN = MQTT_RO_SCHEMA.extend(
+    {
+        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+        vol.Optional(CONF_COMMAND_TOPIC): valid_publish_topic,
+        vol.Optional(CONF_LATEST_VERSION_TEMPLATE): cv.template,
+        vol.Required(CONF_LATEST_VERSION_TOPIC): valid_subscribe_topic,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_PAYLOAD_INSTALL): cv.string,
+        vol.Optional(CONF_RETAIN, default=DEFAULT_RETAIN): cv.boolean,
+    },
+).extend(MQTT_ENTITY_COMMON_SCHEMA.schema)
+
+
+DISCOVERY_SCHEMA = vol.All(PLATFORM_SCHEMA_MODERN.extend({}, extra=vol.REMOVE_EXTRA))
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT update through configuration.yaml and dynamically through MQTT discovery."""
+    setup = functools.partial(
+        _async_setup_entity, hass, async_add_entities, config_entry=config_entry
+    )
+    await async_setup_entry_helper(hass, update.DOMAIN, setup, DISCOVERY_SCHEMA)
+
+
+async def _async_setup_entity(
+    hass: HomeAssistant,
+    async_add_entities: AddEntitiesCallback,
+    config: ConfigType,
+    config_entry: ConfigEntry,
+    discovery_data: dict[str, Any] | None = None,
+) -> None:
+    """Set up the MQTT update."""
+    async_add_entities([MqttUpdate(hass, config, config_entry, discovery_data)])
+
+
+class MqttUpdate(MqttEntity, UpdateEntity, RestoreEntity):
+    """representation of an MQTT update."""
```

Comment:
@emontnemery yes, Shelly should work with this PR for users which are using it via MQTT.

Reply:
@emontnemery My goal was to be able to create `update` entities for Shelly devices using MQTT. Initially, I also thought that one topic and template would be the best solution. However, It turned out that for Shelly Gen2 devices it will not work because these devices send `installed_version` and `latest_version` in separate topics (`progress` is also sent in a separate topic).
In my opinion, separate topics and templates are definitely a more flexible solution. If we decide on one topic and a static payload scheme, we will close access to this platform to devices/services that are not DIY and whose firmware is not easy to change.

Reply:
Perhaps it is an idea to have a optional state topic that can also accept a json dict (like the `siren` platform has), so the state can be overrided and has the capability of providing installed and latest version.
But for this PR, may be it is a good idea to have seperate topic's as it appears this is also a pattern in real live.

Reply:
I thought of one more possibility. `state_topic` can be used for `title`, `entity_picture`, `release_url`, `release_summary` properties, and for `installed_version` and `latest_version` will be separate topics.

Reply:
I think optional the `value_template` could resolve all of them if they are supplied, May be for some other cases they are all supplies in one tipic. The `state_topic` accepted payload could be JSON only. The separate topics are should then be optional probably.

Reply:
@jbouwh I'm not sure I understand you. You mean `state_topic` with JSON data, e.g.:
`{"installed_version": "2.0", "latest_version": "2.1", "title" "Test", "release_url": "https://test.com/", "release_summary": "summary"}`
and optionally `installed_version_topic` and/or `latest_version_topic` if required by the device? If `installed_version_topic`/`latest_version_topic` are present in the config we use them, otherwise we use values from `state_topic` JSON.

Reply:
> @jbouwh I'm not sure I understand you. You mean `state_topic` with JSON data, e.g.: `{"installed_version": "2.0", "latest_version": "2.1", "title" "Test", "release_url": "https://test.com/", "release_summary": "summary"}` and optionally `installed_version_topic` and/or `latest_version_topic` if required by the device? If `installed_version_topic`/`latest_version_topic` are present in the config we use them, otherwise we use values from `state_topic` JSON.

That is indeed what I mean, `state_topic` should be optional in that case (I would advice base the config schema on `MQTT_BASE_SCHEMA`). My idea is that `state_topic` could be used to receive all config options.

Reply:
Thanks @bieniu for explaining 👍